mp4/mdat.go

```diff
@@ -84,3 +92,51 @@ func (m *MdatBox) HeaderSize() uint64 {
 func (m *MdatBox) PayloadAbsoluteOffset() uint64 {
 	return m.StartPos + m.HeaderSize()
 }
+
+// setReadSeeker - set readseeker to read Mdat data.
+// When a file is decoded lazily, the Mdat Data byte slice is nil
+// and this readseeker is to read data whenever the data is needed.
+func (m *MdatBox) setReadSeeker(rs io.ReadSeeker) {
+	m.readSeeker = rs
+}
+
+// ReadData reads Mdat data specified by the start and size.
+// Input argument start is the postion relative to the start of a file.
+func (m *MdatBox) ReadData(start, size int64) ([]byte, error) {
+	// The Mdat box was decoded lazily
+	if m.decLazyDataSize > 0 {
+		if m.readSeeker == nil {
+			return nil, errors.New("lazy mdat mode - expects non-nil readseeker to read data")
+		}
+
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		_, err := m.readSeeker.Seek(start, io.SeekStart)
+		if err != nil {
+			return nil, fmt.Errorf("lazy mdat mode - unable to seek to %d", start)
+		}
+
+		buf := make([]byte, size)
+		n, err := m.readSeeker.Read(buf)
+		if err != nil {
+			return nil, err
+		}
+		if int64(n) != size {
+			return nil, fmt.Errorf("lazy mdat mode - expects to read %d bytes, but only read %d bytes", size, n)
+		}
+		return buf, nil
+	}
+
+	// Otherwise, all Mdat data is in memory
+	mdatPayloadStart := m.PayloadAbsoluteOffset()
+	offsetInMdatData := uint64(start) - mdatPayloadStart
+	endIndexInMdatData := offsetInMdatData + uint64(size)
+
+	// validate if indexes are valid to avoid panics
+	if offsetInMdatData >= uint64(len(m.Data)) || endIndexInMdatData >= uint64(len(m.Data)) {
```

Comment:
endIndexInMdataData >= offsetInMdatData if size >= 0 so you don't need the first part of the test.
On the other hand, there is no check that start > mdatPayloadStart, so it may be a good idea to add that.